src/sdk/pynni/nni/compression/torch/builtin_pruners.py

```diff
@@ -206,25 +259,141 @@ def _get_distance_sum(self, weight, in_idx, out_idx):
             between this specified filter and all other filters.
         in_idx: int
             input channel index of specified filter
-
         Returns
         -------
         float32
             The total distance
         """
         logger.debug('weight size: %s', weight.size())
-        if len(weight.size()) == 4: # Conv2d
+        if len(weight.size()) == 4:  # Conv2d
             w = weight.view(-1, weight.size(-2), weight.size(-1))
             anchor_w = weight[out_idx, in_idx].unsqueeze(0).expand(w.size(0), w.size(1), w.size(2))
-        elif len(weight.size()) == 3: # Conv1d
+        elif len(weight.size()) == 3:  # Conv1d
             w = weight.view(-1, weight.size(-1))
             anchor_w = weight[out_idx, in_idx].unsqueeze(0).expand(w.size(0), w.size(1))
         else:
             raise RuntimeError('unsupported layer type')
         x = w - anchor_w
-        x = (x*x).sum((-2, -1))
+        x = (x * x).sum((-2, -1))
         x = torch.sqrt(x)
         return x.sum()
 
     def update_epoch(self, epoch):
         self.epoch_pruned_layers = set()
+
+
+class FilterPruner(Pruner):
+    """
+    A structured pruning algorithm that prunes the filters of smallest magnitude
+    weights sum in the convolution layers to achieve a preset level of network sparsity.
+    Hao Li, Asim Kadav, Igor Durdanovic, Hanan Samet and Hans Peter Graf,
+    "PRUNING FILTERS FOR EFFICIENT CONVNETS", 2017 ICLR
+    https://arxiv.org/abs/1608.08710
+    """
+
+    def __init__(self, model, config_list):
+        """
+        Parameters
+        ----------
+        model : torch.nn.module
+            Model to be pruned
+        config_list : list
+            List on pruning configs
+        """
+
+        super().__init__(model, config_list)
+        self.if_init_list = {}
+
+    def calc_mask(self, layer, config):
+        """
+        Calculate the mask of given layer.
+        Filters with the smallest sum of its absolute kernel weights are masked.
+        Parameters
+        ----------
+        layer : LayerInfo
+            the layer to instrument the compression operation
+        config : dict
+            layer's pruning config
+        Returns
+        -------
+        torch.Tensor
+            mask of the layer's weight
+        """
+
+        weight = layer.module.weight.data
+        op_name = layer.name
+        op_type = layer.type
+        assert op_type == 'Conv2d', 'FilterPruner only supports 2d convolution layer pruning'
+        if self.if_init_list.get(op_name, True):
+            kernels = weight.shape[0]
```

Comment:
the weight.shape[0]  is number of output channels,   are you intented to use kernels represent number of kernels?

Reply:
Yes this is exactly what I mean.

Reply:
I don't understand your response.  I mean that the number of kernels is `weight.shape[0] * weight.shape[1]`,  if you do not change,  the variable `kernels` is actually the number of output channels, not number of kernels.

Reply:
Yes it should be `out_channels` or `filters`, `kernels` is misleading and I've renamed it.

Reply:
Per the document in this PR, the filters are pruned based on number of filters instead of number of output channels.

Reply:
The number of filters and the number of output channels is actually the same, as per output channel is produced by a corresponding filter.

Reply:
Yes, your are right, I misunderstood it.